### PR TITLE
Update benchmark suite for poliastro v0.14

### PR DESCRIPTION
**Pull request by samtx**
This updates the existing benchmark suite to work with the new poliastro v0.14 APIs and uses flit for build and install.

This is still a work in progress.

### CODE REVIEW

File: asv.conf.json [hunk 1]
```diff
@@ -10,10 +12,43 @@
     // list indicates to just test against the default (latest)
     // version.
     "matrix": {
-        "numpy": ["1.14"],
-        "numba": ["0.39"]
+        "numpy": [],
+        "numba": [],
+        "astropy": [],
```
Comment (astrojuanlu) [inline asv.conf.json]: Do we need to specify all the dependencies here?

File: asv.conf.json [hunk 2]
```diff
@@ -10,10 +12,43 @@
     // list indicates to just test against the default (latest)
     // version.
     "matrix": {
-        "numpy": ["1.14"],
-        "numba": ["0.39"]
+        "numpy": [],
```
Comment (astrojuanlu) [inline asv.conf.json]: Perhaps we should update the versions to something more recent, for example NumPy 1.16 and numba 0.46, but keep the versions anyway. This way we can control for performance changes when the dependencies upgrade.

**Review by astrojuanlu — COMMENTED** [review]
Left a couple of comments, although I see this is still a work in progress. Thanks @samtx for this pull request! It's really needed.

File: benchmarks/perturbations.py [hunk 3]
```diff
@@ -67,5 +69,6 @@ def setup(self):
         self.rho0 = Earth.rho0.to(u.kg / u.km**3).value  # kg/km^3
         self.H0 = Earth.H0.to(u.km).value
         self.tof = 60  # s
-    def time_atmospheric_drag(self):
-        cowell(self.orbit, self.tof, ad=atmospheric_drag, R=self.R, C_D=self.C_D, A=self.A, m=self.m, H0=self.H0, rho0=self.rho0)
+        
+    def time_atmospheric_drag_exponential(self):
+        cowell(self.orbit, self.tof, ad=atmospheric_drag_exponential, R=self.R, C_D=self.C_D, A=self.A, m=self.m, H0=self.H0, rho0=self.rho0)
```
Comment (astrojuanlu) [inline benchmarks/perturbations.py]: Oops, this API changed as well. This should be:

```suggestion
        cowell(self.orbit.attractor.k, self.orbit.r, self.orbit.v, self.tof.reshape(-1), ad=atmospheric_drag_exponential, R=self.R, C_D=self.C_D, A=self.A, m=self.m, H0=self.H0, rho0=self.rho0)
```

and the same for all `cowell` calls.

File: benchmarks/perturbations.py [hunk 4]
```diff
@@ -67,5 +69,6 @@ def setup(self):
         self.rho0 = Earth.rho0.to(u.kg / u.km**3).value  # kg/km^3
```
Comment (astrojuanlu) [inline benchmarks/perturbations.py]:
```suggestion
        self.rho0 = rho0_earth.to_value(u.kg / u.km ** 3)
```

and this comes from `from poliastro.constants import rho0_earth`

**Comment by samtx** [comment]
There are still a few errors:

```
[ 80.00%]
··· perturbations.AtmosphericDrag.time_atmospheric_drag_exponential                                     failed[ 80.00%] ···· Traceback (most recent call last):
                 File "/home/sam/poliastro_benchmarks/venv/lib/python3.8/site-packages/asv/benchmark.py", line 1184, in main_run_server
                   main_run(run_args)
                 File "/home/sam/poliastro_benchmarks/venv/lib/python3.8/site-packages/asv/benchmark.py", line 1052, in main_run
                   skip = benchmark.do_setup()
                 File "/home/sam/poliastro_benchmarks/venv/lib/python3.8/site-packages/asv/benchmark.py", line 583, in do_setup
                   result = Benchmark.do_setup(self)
                 File "/home/sam/poliastro_benchmarks/venv/lib/python3.8/site-packages/asv/benchmark.py", line 515, in do_setup
                   setup(*self._current_params)
                 File "/home/sam/poliastro_benchmarks/benchmarks/perturbations.py", line 71, in setup
                   self.rho0 = rho0_earth.to(u.kg / u.km ** 3)  # kg/km^3
               AttributeError: 'SolarSystemPlanet' object has no attribute 'H0'

[ 90.00%] ··· perturbations.J2_Propagation.time_J2_propagation                                                    failed[ 90.00%] ···· Traceback (most recent call last):
                 File "/home/sam/poliastro_benchmarks/venv/lib/python3.8/site-packages/asv/benchmark.py", line 1184, in main_run_server
                   main_run(run_args)
                 File "/home/sam/poliastro_benchmarks/venv/lib/python3.8/site-packages/asv/benchmark.py", line 1058, in main_run
                   result = benchmark.do_run()
                 File "/home/sam/poliastro_benchmarks/venv/lib/python3.8/site-packages/asv/benchmark.py", line 537, in do_run
                   return self.run(*self._current_params)
                 File "/home/sam/poliastro_benchmarks/venv/lib/python3.8/site-packages/asv/benchmark.py", line 627, in run
                   samples, number = self.benchmark_timing(timer, min_repeat, max_repeat,
                 File "/home/sam/poliastro_benchmarks/venv/lib/python3.8/site-packages/asv/benchmark.py", line 663, in benchmark_timing
                   timing = timer.timeit(number)
                 File "/usr/lib/python3.8/timeit.py", line 176, in timeit
                   timing = self.inner(it, self.timer)
                 File "<timeit-src>", line 6, in inner
                 File "/home/sam/poliastro_benchmarks/benchmarks/perturbations.py", line 31, in time_J2_propagation
                   def time_J2_propagation(self):
               TypeError: cowell() missing 2 required positional arguments: 'v' and 'tofs'

[100.00%] ··· perturbations.J3_Propagation.time_J3_propagation                                                    failed[100.00%] ···· Traceback (most recent call last):
                 File "/home/sam/poliastro_benchmarks/venv/lib/python3.8/site-packages/asv/benchmark.py", line 1184, in main_run_server
                   main_run(run_args)
                 File "/home/sam/poliastro_benchmarks/venv/lib/python3.8/site-packages/asv/benchmark.py", line 1058, in main_run
                   result = benchmark.do_run()
                 File "/home/sam/poliastro_benchmarks/venv/lib/python3.8/site-packages/asv/benchmark.py", line 537, in do_run
                   return self.run(*self._current_params)
                 File "/home/sam/poliastro_benchmarks/venv/lib/python3.8/site-packages/asv/benchmark.py", line 627, in run
                   samples, number = self.benchmark_timing(timer, min_repeat, max_repeat,
                 File "/home/sam/poliastro_benchmarks/venv/lib/python3.8/site-packages/asv/benchmark.py", line 663, in benchmark_timing
                   timing = timer.timeit(number)
                 File "/usr/lib/python3.8/timeit.py", line 176, in timeit
                   timing = self.inner(it, self.timer)
                 File "<timeit-src>", line 6, in inner
                 File "/home/sam/poliastro_benchmarks/benchmarks/perturbations.py", line 52, in time_J3_propagation
                   def time_J3_propagation(self):
               TypeError: cowell() missing 2 required positional arguments: 'v' and 'tofs'
```

File: benchmarks/perturbations.py [hunk 5]
```diff
@@ -64,8 +68,9 @@ def setup(self):
         self.B = self.C_D * self.A / self.m
 
         # parameters of the atmosphere
-        self.rho0 = Earth.rho0.to(u.kg / u.km**3).value  # kg/km^3
+        self.rho0 = rho0_earth.to(u.kg / u.km ** 3)  # kg/km^3
         self.H0 = Earth.H0.to(u.km).value
```
Comment (astrojuanlu) [inline benchmarks/perturbations.py]: This is now `from poliastro.constants import H0_earth`

File: benchmarks/perturbations.py [hunk 6]
```diff
@@ -64,8 +68,9 @@ def setup(self):
         self.B = self.C_D * self.A / self.m
 
         # parameters of the atmosphere
-        self.rho0 = Earth.rho0.to(u.kg / u.km**3).value  # kg/km^3
+        self.rho0 = rho0_earth.to(u.kg / u.km ** 3)  # kg/km^3
```
Comment (astrojuanlu) [inline benchmarks/perturbations.py:103]: Don't we need to use the `.value` here? Otherwise it's an Astropy quantity

**Review by astrojuanlu — COMMENTED** [review]
Left two more comments, we're getting closer!

File: asv.conf.json [hunk 7]
```diff
@@ -1,32 +1,65 @@
...
+        "astropy": "",
+        "astroquery": "",
+        "jplephem": "",
+        "pandas": "",
+        "plotly": "",
+        "scipy": "",
+        "matplotlib": "",
     },
 
+    "include": [
```
Comment (astrojuanlu) [inline asv.conf.json:25]: xref: https://asv.readthedocs.io/en/stable/asv.conf.json.html#include

File: asv.conf.json [hunk 8]
```diff
@@ -1,32 +1,65 @@
...
     // The matrix of dependencies to test.  Each key is the name of a
     // package (in PyPI) and the values are version numbers.  An empty
-    // list indicates to just test against the default (latest)
+    // string indicates to just test against the default (latest)
     // version.
     "matrix": {
-        "numpy": ["1.14"],
-        "numba": ["0.39"]
+        "numpy": "",
+        "numba": "",
```
Comment (astrojuanlu) [inline asv.conf.json:15]: What about fixing the versions? That way we control when do we upgrade the benchmarks and notice when a new dependency introduces a performance change. I propose:

```suggestion
        "numpy": ["1.17.5"],
        "numba": ["0.47.0"],
        "llvmlite": ["0.31.0"]
```

these versions are the oldest ones with Python 3.8 wheels, I think.

File: asv.conf.json [hunk 9]
```diff
@@ -1,32 +1,65 @@
...
+        "astropy": "",
```
Comment (astrojuanlu) [inline asv.conf.json:16]: Do we need to explicitly include all the extra dependencies? If we will just use the newest one, we should not need it, am I right?

File: asv.conf.json [hunk 10]
```diff
@@ -1,32 +1,65 @@
...
+    "version": 1,
+
     // The versions of Python to test against
-    "pythons": ["3.6"],
+    "pythons": ["3.8"], // Python 3.7 included below with extra cached_property dependency
 
-    "environment_type": "conda",
-    "conda_channels": ["conda-forge", "defaults"],
+    // "environment_type": "virtualenv",
+    "environment_type": "virtualenv",
```
Comment (astrojuanlu) [inline asv.conf.json]: Remove duplicated line:

```suggestion
    "environment_type": "virtualenv",
```

File: asv.conf.json [hunk 11]
```diff
@@ -1,32 +1,65 @@
 {
+    "version": 1,
+
     // The versions of Python to test against
-    "pythons": ["3.6"],
+    "pythons": ["3.8"], // Python 3.7 included below with extra cached_property dependency
```
Comment (astrojuanlu) [inline asv.conf.json:5]: :+1:

File: benchmarks/kepler_propagation.py [hunk 12]
```diff
@@ -1,17 +1,37 @@
...
+        orbit.propagate(1.0 * u.min, method=self.method[method])
+
+
+if __name__=="__main__":
+    from itertools import product
+    suite = Suite()
+    for param in suite.params:
+        suite.setup()
+        suite.time_propagation(method=method, ecc=ecc)
+
```
Comment (astrojuanlu) [inline benchmarks/kepler_propagation.py]: Remove final line without newline character?

File: benchmarks/perturbations.py [hunk 13]
```diff
@@ -43,29 +45,56 @@ def setup(self):
...
+    suite.setup()
+    suite.time_J2_propagation()
+
+    suite = J3_Propagation()
+    suite.setup()
+    suite.time_J3_propagation()
+
+    suite = AtmosphericDrag()
+    suite.setup()
+    suite.time_atmospheric_drag_exponential()
```
Comment (astrojuanlu) [inline benchmarks/perturbations.py]: Add newline at the end of the file?

**Review by astrojuanlu — CHANGES_REQUESTED** [review]
Left a few style comments, other than that this is good to go. Also, do you think we can move the files back to a `benchmarks/` directory?

File: asv.conf.json (hunk 7 above)
Comment (samtx) [reply asv.conf.json:25]: This is to make asv create only two environments for benchmarking:
- py3.8 with dependencies
- py3.7 with dependencies and cached_property

**Comment by samtx** [comment]
Some more fixes:
- I used `black` to format everything.
- I added `lambert.py` back to the benchmark suite. I had missed it earlier.
- I tried fixing the dependency version of numpy and numba, but kept getting compilation errors when asv tried to install numpy into the virtual environments. Keeping the dependencies as the latest version was the only option that did not cause any issues. There could be a way to fix the dependency version, but I think that it should be the oldest versions required in poliastro's `pyproject.toml`. I don't think asv automatically re-downloads new dependency versions because then the whole benchmark suite would need to be rerun for all applicable commits. So the dependency versions should be fixed to the latest version available at the time the environment is created.
- I've only tested the benchmark suite with python 3.8. I'm currently installing python 3.7 to confirm that it works as well. 

Another thing left to do would be to make the `README.rst` more thorough. The benchmarks themselves look good.

Also, the reason that the parameterized functions in `Lambert` and `KeplerPropagation` are nested in dictionaries is so that the names appear nicely in the charts. If you parameterize using the function handle directly then you get the `repr()` value of the function in the chart legend.

**Comment by astrojuanlu** [comment]
Thanks a lot @samtx! I think the improvements in the README and the question about the versions can be left for future pull requests. Let's merge this already, and strive for incremental improvements. I will actually try to run the benchmarks again with a wide variety of commits in a dedicated machine.

**Review by astrojuanlu — APPROVED** [review]

**Pull request merged**